### PR TITLE
Add support for multi-qubit-gates to CSPLayout

### LINKED ISSUE #7155: CSP ignores 3-or-more-qubit gates
### Information

- **Qiskit Terra version**: main

### What is the current behavior?

Take the following circuit:

```python
from qiskit import QuantumCircuit
qc = QuantumCircuit(3)
qc.cx(0,1)
qc.ccx(2,0,1)
qc.draw()
```
```
q_0: ──■────■──
     ┌─┴─┐┌─┴─┐
q_1: ┤ X ├┤ X ├
     └───┘└─┬─┘
q_2: ───────■──
```

`CSPLayout` should not be able to find a solution with a layout `0- > 1 -> 2`. However:
```python
from qiskit.transpiler import CouplingMap
from qiskit.transpiler.passes import CSPLayout

pass_ = CSPLayout(CouplingMap([(0, 1), (1, 2)]))
result = pass_(qc)
pass_.property_set["CSPLayout_stop_reason"]
```
```
'solution found'
```

### Suggested solutions

There are two options. If `CSPLayout` encounters a 3-or-more-qubit gate, either assumes that this needs complete connection among the involved qubits, or just fails with reason `3-or-more-qubit gate found`.

### PR DESCRIPTION
**Pull request by Silrem**
### Summary
Fixes #7155. Added support for multi-qubit-gates to CSPLayout. CSPLayout now assumes for every 3+-qubit-gate we need a full connection between all involved qubits.

### Details and comments
I've implemented the version suggested in the issue, assuming all qubits involved in the gate, need a full connection. As this is my first issue and I'm not too familiar with the library yet, I'm wondering about a few things:

* Should I add a test case for this fix?
* Do we need to take into account direction? Should we require a bidirectional full connection?
* This approach does improve CSVLayout, but is the assumption actually fair? For Toffoli it holds, but can we say every 3+-qubit-gate does require a full connection? 

- [x] Possibly add testcase
- [x] Add release notes 
- [ ] Check how/if we need to consider direction
- [ ] Check assumption

### CODE REVIEW

**Comment by Silrem** [comment]
Hey guys,
what's the right way for me to get some feedback on this?

**Comment by 1ucian0** [comment]
Removing the [WIP] is the usual way to signal that :-)

**Comment by Silrem** [comment]
very well :D

**Comment by javabster** [comment]
Hi @Silrem thank-you for your contribution and apologies for the slow response. Yes please do add a test case for this and a bugfix release note. You can check out the [Contributing Guidelines](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md) for details on this 😄

**Comment by Silrem** [comment]
Hi @javabster, thanks for your guidance. I've now added a test case and release notes.

File: releasenotes/notes/fix-multi-qubit-gates-in-CSPLayout-1113fcc7951867e6.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+fixes:
+  - |
+    Added support for multi-qubit-gates to CSPLayout. CSPLayout now
+    assumes for every 3+-qubit-gate we need a full connection between
+    all involved qubits. CSPLayout will no longer erroneously create
+    a layout for circuits with multi-qubit-gates if there is no
```
Comment (javabster) [inline releasenotes/notes/fix-multi-qubit-gates-in-CSPLayout-1113fcc7951867e6.yaml:7]:
```suggestion
    Added support for multi-qubit-gates to 
    :class:`~qiskit.transpiler.passes.layout.CSPLayout`. 
    :class:`~qiskit.transpiler.passes.layout.CSPLayout` now
    assumes for every 3+-qubit-gate we need a full connection between
    all involved qubits. The class will no longer erroneously create
    a layout for circuits with multi-qubit-gates if there is no
```

**Comment by javabster** [comment]
Hey @Silrem it looks like you've got a linting error, you can run `tox -eblack` to fix this

**Comment by 1ucian0** [comment]
blacking in 096a4a00d14296ab0336f1002ee8636b843f0800

File: qiskit/transpiler/passes/layout/csp_layout.py [hunk 2]
```diff
@@ -100,10 +101,15 @@ def __init__(
...
         """run the layout method"""
         qubits = dag.qubits
-        cxs = set()
+        needed_connections = set()
 
         for gate in dag.two_qubit_ops():
-            cxs.add((qubits.index(gate.qargs[0]), qubits.index(gate.qargs[1])))
+            needed_connections.add((qubits.index(gate.qargs[0]), qubits.index(gate.qargs[1])))
+
+        for gate in dag.multi_qubit_ops():
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/csp_layout.py:109]: `two_qubits_ops` and `multi_qubits_ops` do very similar things:
```python
        ops = []
        for node in self.op_nodes(include_directives=False):
            if len(node.qargs) >= X:
                ops.append(node)
        return ops
```
So probably it is better to call `op_nodes` and avoid the double iteration on the structure.

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
Good work! I think there is room for a very small optimization.

File: qiskit/transpiler/passes/layout/csp_layout.py (hunk 2 above)
Comment (Silrem) [reply qiskit/transpiler/passes/layout/csp_layout.py]: thx. change made. iterating directly over op_nodes now.

**Comment by 1ucian0** [comment]
> Do we need to take into account direction? Should we require a bidirectional full connection?
Thanks a complicated one. I think is better not to do it for now.

**Comment by Silrem** [comment]
@javabster thanks for your hint. I've learned some new syntax there. committed as you suggested.

File: test/python/transpiler/test_csp_layout.py [hunk 3]
```diff
@@ -288,6 +288,21 @@ def test_seed(self):
 
         self.assertNotEqual(layout_1, layout_2)
 
+    def test_multi_qubit_gates(self):
+        """multi-qubit gates, like Toffoli, need a complete connection among the involved qubits"""
```
Comment (1ucian0) [inline test/python/transpiler/test_csp_layout.py:292]: Can you also add a test that find the solution?

```suggestion
    def test_multi_qubit_gates(self):
        """Toffoli need a complete connection among the involved qubits."""
        ....

    def test_multi_qubit_gates_fail(self):
        """Toffoli need a complete connection among the involved qubits. No solution on incomplete graph."""
```

**Comment by HuangJunye** [comment]
@Silrem Hey, just want to check in to see if you are still working on the PR. Can you please add the tests as suggested by Luciano?

**Comment by Silrem** [comment]
@HuangJunye At the moment, I'm not working on this PR anymore. Over the last months, other duties have taken over, unfortunately. I might be back in the future, but for now, feel free to let someone else complete this if they want.

**Comment by javabster** [comment]
Closing this PR as contributor isn't able to complete right now. We can re-open at a later date if contributor returns 😄

**Pull request closed without merging**